File: plugins/insight-llm/android/src/main/java/com/cosaxo/insight/llm/InsightLLM.kt
// InsightLLM — Kotlin wrapper over MediaPipe Tasks GenAI's
// LlmInference. Mirrors the iOS Swift wrapper one-to-one: open a
// fresh session per generate() call, decode base64 images to
// Bitmaps via BitmapFactory, stream chunks via the SDK's
// generateResponseAsync coroutine.

package com.cosaxo.insight.llm

import android.content.Context
import android.graphics.BitmapFactory
import android.util.Base64
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.genai.llminference.GraphOptions
import com.google.mediapipe.tasks.genai.llminference.LlmInference
import com.google.mediapipe.tasks.genai.llminference.LlmInferenceSession
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

sealed class InsightLLMError(message: String) : RuntimeException(message) {
    class NotLoaded : InsightLLMError("Model has not been loaded; call setModel() first.")
    class ModelNotMultimodal : InsightLLMError("Active model is text-only.")
    class ImageDecodeFailed : InsightLLMError("Could not decode image bytes.")
    class DownloadFailed(reason: String) : InsightLLMError("Download failed: $reason")
}

enum class LLMReadiness(val value: String) {
    UNINITIALIZED("uninitialized"),
    LOADING("loading"),
    READY("ready"),
    ERROR("error"),
}

class InsightLLM(private val context: Context) {

    private var inference: LlmInference? = null
    private var modelPath: String? = null
    private var maxImages: Int = 0
    private var temperature: Float = 0.6f
    private var topK: Int = 40
    @Volatile var readiness: LLMReadiness = LLMReadiness.UNINITIALIZED
        private set

    fun close() {
        inference?.close()
        inference = null
        readiness = LLMReadiness.UNINITIALIZED
    }

    suspend fun setModel(
        path: String,
        temperature: Float,
        topK: Int,
        maxTokens: Int,
        maxImages: Int,
    ) = withContext(Dispatchers.IO) {
        if (modelPath == path && inference != null) return@withContext
        readiness = LLMReadiness.LOADING

        // Tear down the previous engine before swapping. MediaPipe
        // hangs on to native buffers / GPU resources that aren't
        // GC'd; an explicit close is the only safe way to reload.
        inference?.close()
        inference = null

        val opts = LlmInference.LlmInferenceOptions.builder()
            .setModelPath(path)
            .setMaxTokens(maxTokens)
            // maxNumImages sets the per-session image cap. Zero
            // forces text-only behaviour on a multimodal model.
            .setMaxNumImages(maxImages)
            .build()
        try {
            inference = LlmInference.createFromOptions(context, opts)
            modelPath = path
            this@InsightLLM.maxImages = maxImages
            this@InsightLLM.temperature = temperature
            this@InsightLLM.topK = topK
            readiness = LLMReadiness.READY
        } catch (t: Throwable) {
            readiness = LLMReadiness.ERROR
            throw t
        }
    }

    suspend fun generate(
        prompt: String,
        imageBase64: String?,
        onChunk: (text: String, done: Boolean) -> Unit,
    ): String = withContext(Dispatchers.IO) {
        val engine = inference ?: throw InsightLLMError.NotLoaded()
        val hasImage = !imageBase64.isNullOrEmpty()
        if (hasImage && maxImages == 0) {
            throw InsightLLMError.ModelNotMultimodal()
        }

        // Fresh session per call so previous prompts can't leak.
        val sessionOpts = LlmInferenceSession.LlmInferenceSessionOptions.builder()
            .setTemperature(temperature)
            .setTopK(topK)
            .setGraphOptions(
                GraphOptions.builder()
                    .setEnableVisionModality(hasImage)
                    .build()
            )
            .build()
        val session = LlmInferenceSession.createFromOptions(engine, sessionOpts)

        try {
            if (hasImage) {
                val bitmap = decodeBitmap(imageBase64!!)
                    ?: throw InsightLLMError.ImageDecodeFailed()
                val mpImage = BitmapImageBuilder(bitmap).build()
                session.addImage(mpImage)
            }
            session.addQueryChunk(prompt)

            val buffer = StringBuilder()
            // generateResponseAsync returns a ListenableFuture; the
            // streaming variant uses a per-chunk callback. We
            // bridge the callback to a suspending pump via a
            // CompletableDeferred.
            val deferred = kotlinx.coroutines.CompletableDeferred<String>()
            session.generateResponseAsync { partial, done ->
                buffer.append(partial)
                onChunk(partial, done)
                if (done) {
                    deferred.complete(buffer.toString().trim())
                }
            }
            deferred.await()
        } finally {
            session.close()
        }
    }

    suspend fun downloadModel(
        url: String,
        filename: String?,
        onProgress: (Double) -> Unit,
    ): String = withContext(Dispatchers.IO) {
        val name = filename ?: url.substringAfterLast('/').ifEmpty { "model.task" }
        val dest = File(context.filesDir, name)
        if (dest.exists() && dest.length() > 0) {
            // Already cached. Skip re-download; callers can rename
            // the file or pass a different `filename` to force a
            // fresh pull.
            onProgress(1.0)
            return@withContext dest.absolutePath
        }

        val conn = URL(url).openConnection() as HttpURLConnection
        conn.requestMethod = "GET"
        conn.connect()
        if (conn.responseCode !in 200..299) {
            throw InsightLLMError.DownloadFailed("HTTP ${conn.responseCode}")
        }

        val total = conn.contentLengthLong.takeIf { it > 0 }
        var downloaded = 0L
        val tmp = File(context.filesDir, "$name.part")
        conn.inputStream.use { input ->
            tmp.outputStream().use { out ->
                val buf = ByteArray(64 * 1024)
                while (true) {
                    val n = input.read(buf)
                    if (n <= 0) break
                    out.write(buf, 0, n)
                    downloaded += n
                    if (total != null) {
                        onProgress(downloaded.toDouble() / total.toDouble())
                    }
                }
            }
        }
        if (!tmp.renameTo(dest)) {
            throw InsightLLMError.DownloadFailed("rename to ${dest.name} failed")
        }
        onProgress(1.0)
        dest.absolutePath
    }

    private fun decodeBitmap(base64: String): android.graphics.Bitmap? {
        // Tolerate "data:image/...;base64," prefixes coming from
        // <input type=file> or canvas exports.
        val cleaned = base64.substringAfter(',', base64)
        return try {
            val bytes = Base64.decode(cleaned, Base64.DEFAULT)
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        } catch (t: Throwable) {
            null
        }
    }
}

File: plugins/insight-llm/ios/Sources/InsightLLMPlugin/InsightLLM.swift
// InsightLLM — thin Swift wrapper over MediaPipeTasksGenai.
//
// Lifecycle:
//   setModel(...)  ─→ build LlmInference + an LlmInferenceSession
//   generate(...)  ─→ optionally addImage(), then generateResponseAsync()
//                     and accumulate streamed chunks
//   reset()        ─→ create a fresh session (clears previous turn)
//
// We open a fresh session per generate() call so the model never
// carries unintended state across journal taps. Cheap on Gemma 3n —
// session creation is ~10 ms after the model is already in RAM.

import Foundation
import MediaPipeTasksGenAI
#if canImport(UIKit)
import UIKit
#endif

enum InsightLLMError: Error {
    case notLoaded
    case modelNotMultimodal
    case imageDecodeFailed
    case downloadFailed(String)
}

enum LLMReadiness: String {
    case uninitialized
    case loading
    case ready
    case error
}

final class InsightLLM {
    private var inference: LlmInference?
    private var modelPath: String?
    private var config: LlmInference.Options?
    private(set) var readiness: LLMReadiness = .uninitialized
    private let queue = DispatchQueue(label: "com.cosaxo.insight.llm", qos: .userInitiated)

    func setModel(
        path: String,
        temperature: Float,
        topK: Int,
        maxTokens: Int,
        maxImages: Int
    ) async throws {
        if modelPath == path && inference != nil { return }
        readiness = .loading

        // Tear down any previous engine before constructing the new
        // one — MediaPipe holds GPU buffers and reloading without a
        // close() leaks memory.
        inference = nil

        let options = LlmInference.Options(modelPath: path)
        options.maxTokens = maxTokens
        // maxNumImages is the session-level cap on attached images.
        // Zero forces text-only behaviour even on a multimodal model.
        options.maxNumImages = maxImages

        do {
            let engine = try LlmInference(options: options)
            self.inference = engine
            self.modelPath = path
            self.config = options
            readiness = .ready
            // Stash sampling params on the actor instance; they're
            // applied per-session in generate() since LlmInference
            // itself only takes structural options at init.
            self.temperature = temperature
            self.topK = topK
        } catch {
            readiness = .error
            throw error
        }
    }

    private var temperature: Float = 0.6
    private var topK: Int = 40

    func generate(
        prompt: String,
        imageBase64: String?,
        onChunk: @escaping (_ text: String, _ done: Bool) -> Void
    ) async throws -> String {
        guard let inference = self.inference else {
            throw InsightLLMError.notLoaded
        }
        let hasImage = (imageBase64 != nil && !(imageBase64?.isEmpty ?? true))
        if hasImage && (config?.maxNumImages ?? 0) == 0 {
            throw InsightLLMError.modelNotMultimodal
        }

        // Build a fresh session per call so we never accidentally
        // carry over a previous user's prompt.
        let sessionOpts = LlmInference.Session.Options()
        sessionOpts.temperature = temperature
        sessionOpts.topK = topK
        let session = try LlmInference.Session(
            llmInference: inference,
            options: sessionOpts
        )

        if let base64 = imageBase64, let image = decodeImage(base64: base64) {
            try session.addImage(image: image)
        } else if hasImage {
            throw InsightLLMError.imageDecodeFailed
        }

        try session.addQueryChunk(inputText: prompt)

        // generateResponseAsync streams chunks via the AsyncStream
        // returned from the SDK. We accumulate and forward.
        var buffer = ""
        let stream = session.generateResponseAsync()
        for try await chunk in stream {
            buffer += chunk
            onChunk(chunk, false)
        }
        onChunk("", true)
        return buffer.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func decodeImage(base64: String) -> MPImage? {
        // Tolerate a "data:image/...;base64," prefix.
        let cleaned: String
        if let comma = base64.firstIndex(of: ",") {
            cleaned = String(base64[base64.index(after: comma)...])
        } else {
            cleaned = base64
        }
        guard let data = Data(base64Encoded: cleaned, options: .ignoreUnknownCharacters) else {
            return nil
        }
        #if canImport(UIKit)
        guard let ui = UIImage(data: data) else { return nil }
        return try? MPImage(uiImage: ui)
        #else
        return nil
        #endif
    }

    func downloadModel(
        from url: URL,
        filename: String?,
        onProgress: @escaping (Double) -> Void
    ) async throws -> String {
        let fm = FileManager.default
        let docs = try fm.url(
            for: .documentDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let name = filename ?? url.lastPathComponent
        let dest = docs.appendingPathComponent(name)

        if fm.fileExists(atPath: dest.path) {
            // Already on disk; treat as fully complete and skip the
            // network round trip. Callers can pass a different
            // filename to force a re-download.
            onProgress(1.0)
            return dest.path
        }

        // URLSession download with progress observation. We use a
        // delegate-based session so we get per-byte progress; the
        // simpler `data(from:)` API returns only after the full
        // payload has arrived which is no good for a 2 GB download.
        let delegate = DownloadDelegate(onProgress: onProgress)
        let config = URLSessionConfiguration.default
        let session = URLSession(
            configuration: config,
            delegate: delegate,
            delegateQueue: nil
        )
        let (tmpURL, response) = try await session.download(from: url)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw InsightLLMError.downloadFailed("HTTP \(code)")
        }
        try fm.moveItem(at: tmpURL, to: dest)
        onProgress(1.0)
        return dest.path
    }
}

// URLSessionDownloadDelegate that converts byte-progress callbacks
// into a single Double in [0, 1] for the plugin's notifyListeners
// pump. Held weakly on the session so it deinits when the download
// completes.
private final class DownloadDelegate: NSObject, URLSessionDownloadDelegate {
    let onProgress: (Double) -> Void
    init(onProgress: @escaping (Double) -> Void) {
        self.onProgress = onProgress
    }

    func urlSession(
        _ session: URLSession,
        downloadTask: URLSessionDownloadTask,
        didWriteData bytesWritten: Int64,
        totalBytesWritten: Int64,
        totalBytesExpectedToWrite: Int64
    ) {
        guard totalBytesExpectedToWrite > 0 else { return }
        let pct = Double(totalBytesWritten) / Double(totalBytesExpectedToWrite)
        onProgress(pct)
    }

    func urlSession(
        _ session: URLSession,
        downloadTask: URLSessionDownloadTask,
        didFinishDownloadingTo location: URL
    ) {
        // The plugin moves the file out of the temp location itself
        // (via the await on `session.download(from:)`), so this
        // hook is a no-op for our pump.
    }
}
